File: web3-libs/mod3/playground/aula2/balance_of_erc20.py
# 1. Instale e importe a biblioteca web3.py
from web3 import Web3
import readline

# 2. Configure o provider
RPC_URL = "http://127.0.0.1:8545"
web3 = Web3(Web3.HTTPProvider(RPC_URL))

# 3. Defina o ABI do contrato ERC20
dai_abi = [
    {
        "constant": True,
        "inputs": [{"name": "_owner", "type": "address"}],
        "name": "balanceOf",
        "outputs": [{"name": "balance", "type": "uint256"}],
        "payable": False,
        "stateMutability": "view",
        "type": "function",
    }
]

# Função para solicitar o endereço do contrato
def ask_for_contract_address():
    contract_address = input("Digite o endereço do contrato: ")
    ask_for_account_address(contract_address)

# Função para solicitar o endereço da conta
def ask_for_account_address(contract_address):
    account_address = input("Digite o endereço da conta: ")
    check_balance(contract_address, account_address)

# Função para verificar o saldo
def check_balance(contract_address, account_address):
    dai_contract = web3.eth.contract(address=contract_address, abi=dai_abi)

    try:
        balance = dai_contract.functions.balanceOf(account_address).call()
        print("Saldo do Token:", web3.from_wei(balance, "ether"))
    except Exception as error:
        print("Erro ao consultar o saldo de DAI:", error)

# Inicia o processo
ask_for_contract_address() 

File: web3-libs/mod4/playground/aula2/balanceOfERC20.js
// 1. Instale e importe a biblioteca Ethers.js
import { ethers } from "ethers";
import readline from "readline";

// 2. Configure o provider
const RPC_URL = "http://127.0.0.1:8545";
const provider = new ethers.JsonRpcProvider(RPC_URL);
const daiAbi = [
  {
    constant: true,
    inputs: [{ name: "_owner", type: "address" }],
    name: "balanceOf",
    outputs: [{ name: "balance", type: "uint256" }],
    type: "function",
  },
];

// 3. Obtenha o endereço do contrato e da conta via readline input
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

// Função para solicitar o endereço do contrato
const askForContractAddress = () => {
  rl.question("Digite o endereço do contrato: ", (contractAddress) => {
    askForAccountAddress(contractAddress);
  });
};

// Função para solicitar o endereço da conta
const askForAccountAddress = (contractAddress) => {
  rl.question("Digite o endereço da conta: ", (accountAddress) => {
    checkBalance(contractAddress, accountAddress);
  });
};

// Função para verificar o saldo
const checkBalance = (contractAddress, accountAddress) => {
  const daiContract = new ethers.Contract(contractAddress, daiAbi, provider);

  daiContract.balanceOf(accountAddress)
    .then((balance) => {
      console.log("Saldo do Token:", ethers.formatEther(balance));
    })
    .catch((error) => {
      console.error("Erro ao consultar o saldo:", error);
    })
    .finally(() => {
      rl.close();
    });
};

// Inicia o processo
askForContractAddress();
